Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -265,434 +237,255 @@ final class ConnectionPool {
+            assert(self.waiters.isEmpty)
+            assert(self.availableConnections.isEmpty)
+            assert(self.openedConnectionsCount == 0)
         }
 
         var description: String {
-            return "HTTP1ConnectionProvider { key: \(self.key), state: \(self.state) }"
+            return "HTTP1ConnectionProvider { key: \(self.key) }"
         }
 
-        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
-            self.activityPrecondition(expected: [.opened])
-            let action = self.stateLock.withLock { self.state.connectionAction(for: preference) }
-            switch action {
-            case .leaseConnection(let connection):
-                return connection.removeIdleConnectionHandlersForLease().flatMapError { _ in
-                    connection.closeFuture.flatMap { // We ensure close actions are run first
-                        let defaultEventLoop = self.stateLock.withLock {
-                            self.state.defaultEventLoop
-                        }
-                        return self.makeConnection(on: preference.bestEventLoop ?? defaultEventLoop)
-                    }
-                }
-            case .makeConnection(let eventLoop):
-                return self.makeConnection(on: eventLoop)
-            case .leaseFutureConnection(let futureConnection):
-                return futureConnection
+        var isActive: Bool {
+            return self.lock.withLock {
+                self.active
             }
         }
 
-        func release(connection: Connection) {
-            self.activityPrecondition(expected: [.opened, .closing])
-            let action = self.parentPool.connectionProvidersLock.withLock {
-                self.stateLock.withLock { self.state.releaseAction(for: connection) }
-            }
+        func execute(_ action: Action) {
             switch action {
-            case .succeed(let promise):
-                promise.succeed(connection)
-
-            case .makeConnectionAndComplete(let eventLoop, let promise):
-                self.makeConnection(on: eventLoop).cascade(to: promise)
-
-            case .replaceConnection(let eventLoop, let promise):
-                connection.close().flatMap {
-                    self.makeConnection(on: eventLoop)
-                }.whenComplete { result in
-                    switch result {
-                    case .success(let connection):
-                        promise.succeed(connection)
-                    case .failure(let error):
-                        promise.fail(error)
+            case .lease(let connection, let waiter):
+                // check if we can vend this connection to caller
+                connection.cancelIdleTimeout().flatMapError { error in
+                    // if connection is already inactive, we create a new one.
+                    if error is Connection.InactiveChannelError {
+                        self.openedConnectionsCount += 1
```

Comment:
This is missing the lock, no? Could we have a test case that fails under TSan, showing that this isn’t right?

Reply:
good catch, I think so, yes

Reply:
done